### ISSUE TITLE
Parsing failure on overloaded method in JCLStrings.pas

### ISSUE THREAD

**Issue by Scharrels** [integrated-application-development/sonar-delphi#405]
### Prerequisites

- [x] This bug is in SonarDelphi, not SonarQube or my Delphi code.
- [x] This bug has not already been reported.

### SonarDelphi version

1.18.1

### Delphi version

12.2

### Issue description

DelphiLint is reporting parse errors in the JCL library, specifically in JCLStrings.pas. It seems that sonar-delphi can not resolve ambiguity between two method signatures:

```
    function Insert(Index: SizeInt; const Value: string; Count: SizeInt = 1): TJclStringBuilder; overload;
    function Insert(Index: SizeInt; const Value: array of Char): TJclStringBuilder; overload;
```

This code is used in the following spot in JCLStrings.pas:

```
        sb.Insert(0, '[');
```

This code that shows this mistake is available on Github:

https://github.com/project-jedi/jcl/blob/JCL-Latest/jcl/source/common/JclStrings.pas#L5165

The code compiles in Delphi, so it seems that it is possible to overcome this parsing difference.

### Steps to reproduce

* Include JCL in a Delphi project (available on GetIt).
* Analyze a Delphi unit that includes JclStrings.pas in a uses statement (using DelphiLint)
* Notice the following warning in the DelphiLint log:

```
[DEBUG]                     > JclSimpleXml.pas
[DEBUG]                         > JclCharsets.pas
[DEBUG]                         > JclStrings.pas
[DEBUG]                             > JclAnsiStrings.pas
[DEBUG]                             > JclWideStrings.pas
[DEBUG]                         > JclUnicode.pas
[DEBUG]                         > JclUnicode.pas
[DEBUG]                             > JclSysInfo.pas
[DEBUG]                             > JclWideStrings.pas
[INFO] 0/1 files indexed, current file: C:\Repositories\SafetySoftware\src\Lib\RiskCalculation\ConsequenceRiskCalculations.pas
[DEBUG]                             > JclStringConversions.pas
[DEBUG]                             > JclStreams.pas
[DEBUG]                                 > JclMath.pas
[DEBUG]                                 > JclMath.pas
[DEBUG]                                     > Jcl8087.pas
[DEBUG] Name resolution failed on symbol 'Equal' (JclMath line 4687:12)
[DEBUG] Name resolution failed on symbol 'Equal' (JclMath line 4692:16)
[DEBUG] Name resolution failed on symbol 'Equal' (JclMath line 4782:12)
[DEBUG] Name resolution failed on symbol 'Equal' (JclMath line 4787:16)
[DEBUG]                                 > JclSysUtils.pas
[DEBUG]                                     > JclConsole.pas
[DEBUG]                                     > JclStrings.pas
[DEBUG]                                         > JclLogic.pas
[WARN] Ambiguous declarations could not be resolved
[Occurrence] Insert [5165,11] <JclStrings>
[Declaration] Routine Insert, line 465, params = 3 <JclStrings>
[Declaration] Routine Insert, line 467, params = 2 <JclStrings>
[DEBUG] Name resolution failed on symbol 'Insert' (JclStrings line 5165:11)
[DEBUG]                                         > JclAnsiStrings.pas
```

I have added a bit of extra debug context in this log. I think that the warning in the last part is the important part for this bug. I kept the rest in to give a full view of the context in which this error occurred for us.

### Minimal Delphi code exhibiting the issue

The following class emits the same warning:

```
unit BugReport;

interface

type
  TBugReport = class
  private
    procedure Insert(Index: Integer; const Value: string; Count: Integer = 1); overload;
    procedure Insert(Index: Integer; const Value: array of Char); overload;
  public
    procedure Test;
  end;

implementation

{ TBugReport }

procedure TBugReport.Insert(Index: Integer; const Value: string; Count: Integer);
begin

end;

procedure TBugReport.Insert(Index: Integer; const Value: array of Char);
begin

end;

procedure TBugReport.Test;
begin
  Insert(1, '[');
end;

end.
```

**Comment by cirras** [comment]
Hi @Scharrels,

Thanks for your report, I appreciate the level of detail and minimal reproducible case.

This is a known issue (https://github.com/integrated-application-development/sonar-delphi/issues/285) that will be fixed when we add support for constant expressions in the analyzer's semantic model (https://github.com/integrated-application-development/sonar-delphi/issues/116), which will allow us to handle tricky cases like this where an overload was eliminated due to a constant expression being passed as the argument.
